cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerHttpHandlerInternal.java

```diff
@@ -60,26 +64,100 @@
 @Singleton
 @Path(Constants.Gateway.INTERNAL_API_VERSION_3 + "/worker")
 public class TaskWorkerHttpHandlerInternal extends AbstractHttpHandler {
+  private enum RestartState {
+    /**
+     * Pod has not timed out for periodic restart
+     */
+    NOT_TIMED_OUT,
+    /**
+     * Pod has timed out and should restart if there is no inflight request
+     */
+    RESTART_IF_NO_INFLIGHT_REQUEST,
+    /**
+     * Pod should restart now
+     */
+    RESTART_NOW
+  }
+
+  /**
+   * Fraction of duration which will be used for calculating a range.
+   */
+  private static final double DURATION_FRACTION = 0.1;
   private static final Logger LOG = LoggerFactory.getLogger(TaskWorkerHttpHandlerInternal.class);
   private static final Gson GSON = new GsonBuilder().registerTypeAdapter(BasicThrowable.class,
                                                                          new BasicThrowableCodec()).create();
   private final RunnableTaskLauncher runnableTaskLauncher;
   private final BiConsumer<Boolean, String> stopper;
+
   private final AtomicInteger inflightRequests = new AtomicInteger(0);
+
+  /**
+   * Holds the total number of requests that have been executed by this handler that should count toward max allowed.
+   */
+  private final AtomicInteger totalRequestsWithUserCodeExecution = new AtomicInteger(0);
+
   private final String metadataServiceEndpoint;
 
+  private volatile RestartState restartState = RestartState.NOT_TIMED_OUT;
+
   public TaskWorkerHttpHandlerInternal(CConfiguration cConf, Consumer<String> stopper) {
+    int containerKillAfterExecution = cConf.getInt(Constants.TaskWorker.CONTAINER_KILL_AFTER_EXECUTION, 0);
     this.runnableTaskLauncher = new RunnableTaskLauncher(cConf);
     this.metadataServiceEndpoint = cConf.get(Constants.TaskWorker.METADATA_SERVICE_END_POINT);
     this.stopper = (terminate, className) -> {
-      if (cConf.getBoolean(Constants.TaskWorker.CONTAINER_KILL_AFTER_EXECUTION) && terminate) {
-        if (className != null) {
+      synchronized (this) {
+        switch (restartState) {
+          case RESTART_NOW:
+            stopper.accept(className);
+            return;
+          case RESTART_IF_NO_INFLIGHT_REQUEST:
+            if (inflightRequests.compareAndSet(0, 1)) {
+              // there is no ongoing request, and pod needs to restart
+              stopper.accept(className);
+            } else {
+              // there is an ongoing request. Pod will restart once the request finishes.
+              restartState = RestartState.RESTART_NOW;
+            }
+            return;
+        }
```

Comment:
nit: missing NOT_TIMEOUT and default case? even if it is just fall-through for them